### PR TITLE
feat: add ts types

### LINKED ISSUE #29: typescript support
Please consider to officially include the typescript declaration file.
You can copy from my package: https://www.npmjs.com/package/typestub-multibase
repo: https://github.com/beenotung/typestub-multibase

### PR DESCRIPTION
**Pull request by hugomrdias**
needs https://github.com/ipfs/aegir/pull/671

closes #54 
closes #29

check published docs https://multiformats.github.io/js-multibase

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,20 +1,20 @@
...
- * @module Multibase
  */
 'use strict'
 
 const constants = require('./constants')
 const { encodeText, decodeText, concat } = require('./util')
 
-/** @typedef {import("./base")} Base */
+/** @typedef {import('./base')} Base */
+/** @typedef {import("./types").BaseNamesCodes} BaseNamesCodes */
```
Comment (Gozala) [inline src/index.js:11]:
```suggestion
/** @typedef {import("./types").BaseNameOrCode} BaseNameOrCode */
```

File: src/index.js [hunk 2]
```diff
@@ -1,20 +1,20 @@
...
+/** @typedef {import("./types").BaseCodes} BaseCodes */
```
Comment (Gozala) [inline src/index.js:12]:
```suggestion
/** @typedef {import("./types").BaseCode} BaseCode */
```

File: src/constants.js [hunk 3]
```diff
@@ -45,15 +50,17 @@ const constants = [
...
   return prev
-}, {})
+}, /** @type {Record<BaseNames,Base>} */({}))
 
+/** @type {Record<BaseCodes,Base>} */
 const codes = constants.reduce((prev, tupple) => {
   prev[tupple[1]] = names[tupple[0]]
   return prev
-}, {})
+}, /** @type {Record<BaseCodes,Base>} */({}))
```
Comment (Gozala) [inline src/constants.js:63]:
```suggestion
}, /** @type {Record<BaseCode, Base>} */({}))
```

File: src/constants.js [hunk 4]
```diff
@@ -45,15 +50,17 @@ const constants = [
   ['base64urlpad', 'U', rfc4648(6), 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_=']
 ]
 
+/** @type {Record<BaseNames,Base>} */
 const names = constants.reduce((prev, tupple) => {
   prev[tupple[0]] = new Base(tupple[0], tupple[1], tupple[2], tupple[3])
   return prev
-}, {})
+}, /** @type {Record<BaseNames,Base>} */({}))
```
Comment (Gozala) [inline src/constants.js:57]:
```suggestion
}, /** @type {Record<BaseName, Base>} */({}))
```

File: src/constants.js [hunk 5]
```diff
@@ -14,10 +19,10 @@ const identity = () => {
 }
 
 /**
- * @typedef {import('./base').CodecFactory} CodecFactory
  *
  * name, code, implementation, alphabet
- * @type {Array<[string, string, CodecFactory, string]>}
+ *
+ * @type {Array<[BaseNames, BaseCodes, CodecFactory, string]>}
```
Comment (Gozala) [inline src/constants.js:25]:
```suggestion
 * @type {Array<[BaseName, BaseCode, CodecFactory, string]>}
```

File: src/constants.js [hunk 6]
```diff
@@ -1,11 +1,16 @@
...
 
 const baseX = require('@multiformats/base-x')
 const Base = require('./base.js')
-const rfc4648 = require('./rfc4648')
+const { rfc4648 } = require('./rfc4648')
 const { decodeText, encodeText } = require('./util')
 
+/** @typedef {import('./types').CodecFactory} CodecFactory */
+/** @typedef {import('./types').Codec} Codec */
+/** @typedef {import('./types').BaseNames} BaseNames */
```
Comment (Gozala) [inline src/constants.js:10]:
```suggestion
/** @typedef {import('./types').BaseName} BaseName */
```

File: src/constants.js [hunk 7]
```diff
@@ -1,11 +1,16 @@
...
+/** @typedef {import('./types').BaseCodes} BaseCodes */
```
Comment (Gozala) [inline src/constants.js:11]:
```suggestion
/** @typedef {import('./types').BaseCode} BaseCode */
```

**Review by Gozala — CHANGES_REQUESTED** [review]
Mostly nits & suggestions. Only thing I do feel strongly about is type names that I think should be singular instead of plural.

I also flagged out few things with ⚠️ which appear to be breaking changes and might be worth reconsideration.

With type names changed 👍from me and no need for another review cycle.

File: README.md [hunk 8]
```diff
@@ -72,55 +57,6 @@ console.log(decodedBytes.toString())
 ## API
 https://multiformats.github.io/js-multibase/
```
[the remark this answers is not in the document]
Comment (hugomrdias) [reply README.md:58]: Suggestions ?

File: src/util.js [hunk 9]
```diff
@@ -1,6 +1,6 @@
-// @ts-check
 'use strict'
 
+// @ts-ignore
```
[the remark this answers is not in the document]
Comment (hugomrdias) [reply src/util.js:3]: well

error TS2339: Property 'TextDecoder' does not exist on type 'typeof import("/Users/hd/code/pl/js-multibase/node_modules/web-encoding/src/lib")'.

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (Gozala) [reply README.md:58]: I don’t know maybe API Docs badge at the top ?

**Review by Gozala — APPROVED** [review]

**Review by achingbrain — APPROVED** [review]
LGTM, just needs the gh url for aegir removing.

**Pull request merged**